docs/fapi.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "FAPI 1"
.TH FAPI 1 "2014-07-31" "fapi 1.0.2" "User Commands"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
fapi \- A humble command line tool to manage F5 BigIP loadbalancers
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
Just run
.PP
.Vb 1
\&  fapi \-h
.Ve
.PP
or
.PP
.Vb 2
\&  alias f=fapi
\&  f \-h
.Ve
.PP
because it's shorter to type.
.SH "ABOUT"
.IX Header "ABOUT"
This is a simple command line client to do basic stuff with the iControl F5 \s-1API\s0 
such as:
.PP
.Vb 4
\&  Managing Monitors
\&  Managing Nodes
\&  Managing Pools
\&  Managing Virtual Servers
.Ve
.PP
And some extended usage:
.PP
.Vb 4
\&  Managing Folders
\&  Managing Self IPs
\&  Managing Traffic Groups
\&  Managing VLANs
.Ve
.PP
This is a private programming project programmed in my spare time. Therefore I
didn't bother to put it on a public website and github. Please open bug 
reports, feature requests and pull requests at
<https://github.com/rantanplan/fapi>.
.PP
\&\fB\s-1CAUTION:\s0 This script has been tested on Debian GNU/Linux Wheezy only.\fR
.SH "BIGSUDS"
.IX Header "BIGSUDS"
.SS "Requirement of bigsuds"
.IX Subsection "Requirement of bigsuds"
This tool depends on bigsuds. Please install this library from F5 dev central
manually. Otherwise this script will not work.
.PP
You can download bigsuds from here:
.PP
https://devcentral.f5.com/d/bigsuds\-python\-icontrol\-library <https://devcentral.f5.com/d/bigsuds-python-icontrol-library>
.PP
Unzip it and run
.PP
.Vb 1
\&  sudo python setup.py install
.Ve
.PP
You may also install bigsuds from the contrib dir of the fapi source tree.
.SS "iControl reference"
.IX Subsection "iControl reference"
Through bigsuds you can do everything what iControl can do:
.PP
<https://devcentral.f5.com/wiki/icontrol.apireference.ashx>
.SH "QUICK START"
.IX Header "QUICK START"
Update your sources list:
.PP
.Vb 4
\&  curl http://deb.buetow.org/apt/pubkey.gpg | sudo apt\-key add \-
\&  echo deb http://deb.buetow.org/apt wheezy main |
\&    sudo tee /etc/apt/sourcees.list.d/buetoworg.list
\&  sudo aptitude update
.Ve
.PP
And run
.PP
.Vb 3
\&  sudo aptitude install fapi
\&  cp /usr/share/fapi/fapi.conf.sample ~/.fapi.conf
\&  vim ~/.fapi.conf
.Ve
.PP
Or if you want to install it from the source dir, just run:
.PP
.Vb 2
\&  make
\&  sudo make install
.Ve
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.SS "Listing"
.IX Subsection "Listing"
If you want to list all configured objects on your partition just run
.PP
.Vb 3
\&  fapi node # To list all nodes
\&  fapi pool # To list all pool
\&  ... # etc
.Ve
.SS "Setting up a simple pool"
.IX Subsection "Setting up a simple pool"
.Vb 4
\&  # Creating two nodes, fapi auto resolves the IP addresses, and use the
\&  # FQDN as the node name. 
\&  fapi node fooserver1.example.com create
\&  fapi node fooserver2.example.com create
\&
\&  # Creating a pool and add the nodes to it. Also specify the node ports to 
\&  # use by the monitors (and maybe PAT if enabled)
\&  fapi pool foopool create
\&  fapi pool foopool add member fooserver1.example.com:80
\&  fapi pool foopool add member fooserver2.example.com:80
\&
\&  # Add a monitor to the pool
\&  fapi pool foopool add monitor http_lbtest
.Ve
.SS "Setting up a simple nPath Service"
.IX Subsection "Setting up a simple nPath Service"
A simple nPath service can be created as follows.
.PP
.Vb 2
\&  fapi vserver myvserver.example.com:80 create PROTOCOL_TCP nPath
\&  fapi vserver myvserver.example.com:80 set pool foopool
\&
\&  # Restrict the vservers to a specific VLAN (IMPORTANT! security
\&  # hole otherwise!)
\&  fapi vserver myvserver.example.com:80 set vlan VLANNAME
\&  fapi vserver myvserver.example.com:443 set vlan VLANNAME
\&
\&  # Put the VirtualAddress of the vserver into a specific traffic group
\&  fapi vip myserver.example.com set tgroup some\-traffic\-group
.Ve
.PP
And everything can be deleted as folows:
.PP
.Vb 2
\&  # You can also specify the full object name (including the partition)
\&  fapi vserver /Common/myvserver.example.com_80 delete
\&
\&  # Or just the way the service was created from command line
\&  fapi vserver myvserver.example.com:443 delete
\&
\&  fapi pool foopool delete
\&  fapi node fooserver1.example.com delete
\&  fapi node fooserver2.example.com delete
.Ve
.SS "Setting up simple \s-1HTTP\s0 \s-1NAT\s0 Services"
.IX Subsection "Setting up simple HTTP NAT Services"
A simple \s-1HTTP\s0 NATed service can be created as follows.
.PP
.Vb 3
\&  fapi vserver myvserver.example.com:80 create PROTOCOL_TCP http
\&  fapi vserver myvserver.example.com:80 set vlan VLANNAME
\&  fapi vserver myvserver.example.com:80 set pool foopool
.Ve
.PP
In order to make this work your application servers need to have setup a
default route to the loadbalancers floating self \s-1IP\s0.
.SS "Setting up simple \s-1SNAT\s0 Services"
.IX Subsection "Setting up simple SNAT Services"
Same as setting up a NATed services, but you don't need to configure default 
routes from your application servers to the loadbalancers floating self \s-1IP\s0.
.PP
You need also to set the \s-1SNAT\s0 flag as follows:
.PP
.Vb 1
\&  fapi vserver myvserver.example.com:80 set snat automap
.Ve
.SS "Settung up simple \s-1SNAT\s0 Services with \s-1SSL\s0 offloading enabled"
.IX Subsection "Settung up simple SNAT Services with SSL offloading enabled"
Just like \s-1SNAT\s0 service (use port 443) but with these additional steps
.PP

debian/changelog
fapi (1.0.2) unitix-70; urgency=low

  * Fix distribution

 -- Paul Buetow <paul@buetow.org>  Mon, 17 Nov 2014 16:36:55 +0100

fapi (1.0.1) stable; urgency=low

  * Some documentation fixes.
  * Some other fixes as well.

 -- Paul buetow <paul@buetow.org>  Mon, 17 Nov 2014 15:07:55 +0100

fapi (1.0.0) stable; urgency=low

  * Add 'get curl' option to generate test commands to check wether a service
  is available or not anymore.

  * Add possibility to create, delete HA Groups, to assign them to traffic
  groups and to add and remove LACP trunk members.

 -- Paul Buetow <paul@buetow.org>  Wed, 30 Jul 2014 20:44:48 +0200

fapi (0.9.5) stable; urgency=low

  * Fixes the -a argument, didn't really work

 -- Paul Buetow <paul@buetow.org>  Mon, 07 Jul 2014 18:43:09 +0200

fapi (0.9.4) stable; urgency=low

  * Fix fapi.buetow.org HP (Synopsis text)

 -- Paul Buetow <paul@buetow.org>  Wed, 25 Jun 2014 09:25:32 +0200

fapi (0.9.3) stable; urgency=low

  * Fix the extended synopsis

 -- Paul Buetow <paul@buetow.org>  Tue, 24 Jun 2014 19:37:43 +0200

fapi (0.9.2) stable; urgency=low

  * fapi.buetow.org now also shows the extended synopsis

 -- Paul Buetow <paul@buetow.org>  Tue, 24 Jun 2014 19:31:08 +0200

fapi (0.9.1) stable; urgency=low

  * Add -E argument, which shows the extended synopsis 

 -- Paul Buetow <paul@buetow.org>  Tue, 24 Jun 2014 19:28:01 +0200

fapi (0.9.0) stable; urgency=low

  * Can get and set poolmember ratios

 -- Paul Buetow <paul@buetow.org>  Tue, 24 Jun 2014 19:13:51 +0200

fapi (0.8.3) stable; urgency=low

  * fix in docu
  * also no vlan failsave by default

 -- Paul Buetow <paul@buetow.org>  Thu, 19 Jun 2014 08:21:40 +0200

fapi (0.8.2) stable; urgency=low

  * Add Package for Debian Jessie

 -- Paul Buetow <paul@buetow.org>  Sat, 14 Jun 2014 11:00:31 +0200

fapi (0.8.1) stable; urgency=low

  * Hotfix release, could not manage nodes

 -- Paul Buetow <paul@buetow.org>  Fri, 30 May 2014 20:03:31 +0200

fapi (0.8.0) stable; urgency=low

  * Add option -a
  * Add possibility to manage SSL Client Profiles
  * Add possibility to manage VirtualServer rules
  * Add gitignore file
  * Small bugfixes
  * Enhancements in the documentation/manual page

 -- Paul Buetow <paul@buetow.org>  Tue, 27 May 2014 20:17:19 +0200

fapi (0.7.3) stable; urgency=low

  * Fixes documentation

 -- Paul Buetow <paul@buetow.org>  Fri, 16 May 2014 20:13:29 +0200

fapi (0.7.2) stable; urgency=low

  * Fixes documentation

 -- Paul Buetow <paul@buetow.org>  Fri, 16 May 2014 20:11:45 +0200

fapi (0.7.1) stable; urgency=low

  * Add example how to create a NATed VirtualServer...
  * Add example how to create a SNATed VirtualServer...
  ... to the manpage

 -- Paul Buetow <paul@buetow.org>  Fri, 16 May 2014 20:01:48 +0200

fapi (0.7.0) stable; urgency=low

  * Can also set and get the vlan of a vserver
  * Can also set SNAT to automap
  * Small updates in the documentation

 -- Paul Buetow <paul@buetow.org>  Fri, 16 May 2014 17:52:31 +0200

fapi (0.6.2) stable; urgency=low

  * Some more PEP8

 -- Paul Buetow <paul@buetow.org>  Tue, 29 Apr 2014 07:44:58 +0200

fapi (0.6.1) stable; urgency=low

  * Minor fix in synopsis.txt generation

 -- Paul Buetow <paul@buetow.org>  Tue, 29 Apr 2014 07:39:13 +0200

fapi (0.6.0) stable; urgency=low

  * Count version to 0.6.0, it's the 6th release introducing new features
  * Add colorful terminal output (can be disabled with -n)
  * Add initial interactive shell (-i)
  * Respect more of PEP8

 -- Paul Buetow <paul@buetow.org>  Tue, 29 Apr 2014 07:35:52 +0200

fapi (0.0.11) stable; urgency=low

  * Rename partition to folder

 -- Paul Buetow <paul@buetow.org>  Thu, 24 Apr 2014 20:39:15 +0200

fapi (0.0.10) stable; urgency=low

  * Don't include the f command anymore, use an alias to fapi instead

 -- Paul Buetow <paul@buetow.org>  Thu, 24 Apr 2014 20:19:26 +0200

fapi (0.0.9) stable; urgency=low

  * Add synopsis.txt

 -- Paul Buetow <paul@buetow.org>  Thu, 24 Apr 2014 08:59:25 +0200

fapi (0.0.8) stable; urgency=low

  * Can manage folders.

 -- Paul Buetow <paul@buetow.org>  Wed, 23 Apr 2014 22:38:05 +0200

fapi (0.0.7) stable; urgency=low

  * Add the functionality to specify explicitly IP addresses even if the 
    FQDN does not resolve to it. 

 -- Paul Buetow <paul@buetow.org>  Tue, 22 Apr 2014 20:39:28 +0200

fapi (0.0.6) stable; urgency=low

  * Introduce a bunch of new options.

 -- Paul Buetow <paul@buetow.org>  Wed, 16 Apr 2014 22:27:41 +0200

fapi (0.0.5) stable; urgency=low


src/fapi.py
#!/usr/bin/env python

# 2014 (c) Paul C. Buetow

import argparse
import base64
import bigsuds
import getpass 
import pprint
import socket
import sys
import re

from os.path import expanduser
from inspect import isfunction

import ConfigParser

__program__ = 'fapi'
__version__ = 'VERSION_DEVEL' # Replaced by a Makefile tsubet
__prompt__  = '>>>' # Default prompt


def print_synopsis():
    ''' Prints the full Synopsis string '''

    print "\n".join([
        'This is %s version %s' % (__program__, __version__),
        '',
        'Synopsis:',
        ' fapi monitor',
        ' fapi monitor MONITORNAME get desc|state',
        ' fapi node',
        ' fapi node NODENAME create|delete',
        ' fapi node NODENAME get detail|status',
        ' fapi pool',
        ' fapi pool POOLNAME add member MEMBERNAME:PORT',
        ' fapi pool POOLNAME add monitor MONITORNAME',
        ' fapi pool POOLNAME create [LIST,OF,POOL,MEMBERS:PORT]',
        ' fapi pool POOLNAME delete',
        ' fapi pool POOLNAME del member MEMBERNAME:PORT',
        ' fapi pool POOLNAME del monitors',
        ' fapi pool POOLNAME get detail|lbmethod|members|monitor|status',
        ' fapi pool POOLNAME set lbmethod LBMETHOD',
        ' fapi vserver',
        ' fapi vserver VSERVERNAME create [protocol] [profile] [poolname] [mask]',
        ' fapi vserver VSERVERNAME delete',
        ' fapi vserver VSERVERNAME get brief|detail|status',
        ' fapi vserver VSERVERNAME set nat|pat disabled|enabled',
        ' fapi vserver VSERVERNAME set pool POOLNAME',
        ' fapi vserver VSERVERNAME set snat none',
    ])



class Fapi(object):
    ''' The main F5 API Tool Object '''

    def __init__(self, args):
        ''' Initialize the config file, username and password '''

        self._args = args
        self._config = ConfigParser.ConfigParser()
        self._config.read(args.C)
        self._partition = self._config.get('fapi', 'partition')


    def __login(self):
        ''' Logs into the F5 BigIP SOAP API and changes the partition'''

        c = self._config
        a = self._args

        if c.has_option('fapi', 'username'):
            username = c.get('fapi', 'username')
        else:
            username = getpass.getuser()
        if c.has_option('fapi', 'password64'):
            password = base64.decodestring(c.get('fapi', 'password64'))
        else:
            prompt = 'Enter API password for user %s: ' % username
            password = getpass.getpass(prompt)
        self.info('Login to BigIP API with user %s' % username)

        # Try a comma separated lists of F5 boxes, use the first one
        err = None
        for hostname in c.get('fapi', 'hostnames').split(','):
            try:
                self.info('Trying to login to \'%s\'' % hostname)
                self._f5 = bigsuds.BIGIP(hostname = hostname,
                                         username = username,
                                         password = password)
                self._f5.Management.Partition.set_active_partition(self._partition)
                self.info('Set partition to \'%s\'' % self._partition)
                err = None
                break
            except Exception, e:
                err = '%s:%s' % (hostname, e)
                pass

        if err:
            self.info(err)
            sys.exit(2)


    def info(self, message):
        ''' Prints an informational message to stderr '''

        if self._args.v: print >> sys.stderr, '%s %s' % (__prompt__, message)


    def out(self, result):
        ''' Prints an iControl result to stdout '''

        if result != None:
            pp = pprint.PrettyPrinter(indent=4)
            pp.pprint(result)


    def lookup(self, what):
        ''' Does a DNS lookup to fetch the FQDN and all the IPs '''

        tmp = what.split(':')
        if 1 == len(tmp): tmp.append('80')
        what = tmp[0]
        port = tmp[1]
        try:
            data = socket.gethostbyname_ex(what)
        except Exception, e:
            self.info('Can\'t resolve \'%s\': %s' % (what, e))
            sys.exit(2)
        fqdn = data[0]
        ips = data[2]
        if len(ips) > 1:
            self.info('\'%s\' resolves to multiple ips \'%s\'' % (fqdn, ips))
            sys.exit(2)
        return (fqdn, ips[0], port)

    def __do_node(self, f5):
        ''' Do stuff concerning nodes '''

        a = self._args

        if not a.name:
            return lambda: f5().get_list()

        if a.sub == 'get':
            if a.sub2 == 'detail':
                def detail(f5):
                    d = {}
                    d['connection_limit'] = f5().get_connection_limit([a.name])
                    d['default_node_monitor'] = f5().get_default_node_monitor()
                    d['description'] = f5().get_description([a.name])
                    d['dynamic_ratio'] = f5().get_dynamic_ratio_v2([a.name])
                    d['monitor_instance'] = f5().get_monitor_instance([a.name])
                    d['monitor_rule'] = f5().get_monitor_rule([a.name])
                    d['monitor_status'] = f5().get_monitor_status([a.name])
                    d['object_status'] = f5().get_object_status([a.name])
                    d['rate_limit'] = f5().get_rate_limit([a.name])
                    d['ratio'] = f5().get_ratio([a.name])
                    d['session_status'] = f5().get_session_status([a.name])
                    return d
                return lambda: detail(f5)
            if a.sub2 == 'status':
                return lambda: f5().get_monitor_status([a.name])

        elif a.sub == 'create':
                try:
                    data = socket.gethostbyname_ex(a.name)
                except Exception, e:
                    self.info('Can\'t resolve \'%s\': %s' % (a.name, e))
                    sys.exit(2)
                fqdn, ip, _ = self.lookup(a.name)
                return lambda: f5().create([fqdn],[ip],[0])

        elif a.sub == 'delete':
                return lambda: f5().delete_node_address([a.name])


    def __do_monitor(self, f5):
        ''' Do stuff concerning monitor templates '''

        a = self._args

        if not a.name:
            return lambda: f5().get_template_list()

        if a.sub == 'get':
            monitorname = a.sub3
            if a.sub2 == 'desc':
                return lambda: f5().get_description([monitorname])
            if a.sub2 == 'state':
                return lambda: f5().get_template_state([monitorname])


    def __do_pool(self, f5):
        ''' Do stuff concerning pools '''

        a = self._args

        if not a.name:
            return lambda: f5().get_list()

        if a.sub == 'get':
            if a.sub2 == 'detail':
                def detail(f5):
                    d = {}
                    d['allow_nat_state'] = f5().get_allow_nat_state([a.name])
                    d['allow_snat_state'] = f5().get_allow_snat_state([a.name])
                    d['description'] = f5().get_description([a.name])
                    d['lb_method'] = f5().get_lb_method([a.name])
                    d['member'] = f5().get_member_v2([a.name])
                    d['object_status'] = f5().get_object_status([a.name])
                    d['profile'] = f5().get_profile([a.name])
                    return d
                return lambda: detail(f5)
            elif a.sub2 == 'lbmethod':
                return lambda: f5().get_lb_method([a.name])
            elif a.sub2 == 'members':
                return lambda: f5().get_member_v2([a.name])
            elif a.sub2 == 'monitor':
                return lambda: f5().get_monitor_instance([a.name])
            elif a.sub2 == 'status':
                return lambda: f5().get_object_status([a.name])

        elif a.sub == 'set':
            if a.sub2 == 'lbmethod':
                lbmethod = a.sub3
                return lambda: f5().set_lb_method([a.name], [lbmethod])

        elif a.sub == 'create':
                poolmembers = []
                method = a.m
                if a.sub3:
                    for x in a.sub3.split(','):
                        fqdn, ip, port = self.lookup(x)
                        pm = { 'address' : fqdn, 'port' : port }
                        poolmembers.append(pm)
                return lambda: f5().create_v2([a.name],[method],[poolmembers])

        elif a.sub == 'delete':
            return lambda: f5().delete_pool([a.name])

        elif a.sub == 'add':
            if a.sub2 == 'member':
                fqdn, _, port = self.lookup(a.sub3)
                member = [{ 'address' : fqdn, 'port' : port }]
                return lambda: f5().add_member_v2([a.name], [member])
            elif a.sub2 == 'monitor':
                monitorname = a.sub3
                rule = {
                    'type': 'MONITOR_RULE_TYPE_SINGLE',
                    'quorum': long(0),
                    'monitor_templates': [ monitorname ],
                }
                association = { 'pool_name': a.name, 'monitor_rule': rule }
                return lambda: f5().set_monitor_association([association])

        elif a.sub == 'del':
            if a.sub2 == 'member':
                fqdn, _, port = self.lookup(a.sub3)
                member = [{ 'address' : fqdn, 'port' : port }]
                return lambda: f5().remove_member_v2([a.name], [member])
            elif a.sub2 == 'monitors':
                # Removes all monitor associations, not just one
                return lambda: f5().remove_monitor_association([a.name])


    def __do_vserver(self, f5):
        ''' Do stuff concerning virtual servers '''

        a = self._args

        if not a.name:
            return lambda: f5().get_list()

        # Check for Pattern like /partition/foo-bar.example.com_443
        m = re.match('^(.*)_(\d+)$', a.name)
        if m:
          fqdn = m.group(1)
          port = m.group(2)
          _, ip, _ = self.lookup(fqdn)
        else:
          fqdn, ip, port = self.lookup(a.name)

        name = fqdn + '_' + port

        if a.sub == 'get':
            if a.sub2 == 'detail':
                def detail(f5):
                    d = {}
                    d['actual_hardware_acceleration'] = f5().get_actual_hardware_acceleration([name])
                    d['auto_lasthop'] = f5().get_auto_lasthop([name])
                    d['bw_controller_policy'] = f5().get_bw_controller_policy([name])
                    d['clone_pool'] = f5().get_clone_pool([name])
                    d['connection_limit'] = f5().get_connection_limit([name])
                    d['default_pool_name'] = f5().get_default_pool_name([name])
                    d['description'] = f5().get_description([name])
                    d['destination'] = f5().get_destination_v2([name])
                    d['enabled_state'] = f5().get_enabled_state([name])
                    d['fallback_persistence_profile'] = f5().get_fallback_persistence_profile([name])
                    d['gtm_score'] = f5().get_gtm_score([name])
                    d['last_hop_pool'] = f5().get_last_hop_pool([name])
                    d['object_status'] = f5().get_object_status([name])
                    d['persistence_profile'] = f5().get_persistence_profile([name])
                    d['profile'] = f5().get_profile([name])
                    d['protocol'] = f5().get_protocol([name])
                    d['rule'] = f5().get_rule([name])
                    d['snat_pool'] = f5().get_snat_pool([name])
                    d['snat_type'] = f5().get_snat_type([name])
                    d['source_address'] = f5().get_source_address([name])
                    d['source_address_translation_lsn_pool'] = f5().get_source_address_translation_lsn_pool([name])
                    d['source_address_translation_snat_pool'] = f5().get_source_address_translation_snat_pool([name])
                    d['source_address_translation_type'] = f5().get_source_address_translation_type([name])
                    d['source_port_behavior'] = f5().get_source_port_behavior([name])
                    d['translate_address_state'] = f5().get_translate_address_state([name])
                    d['translate_port_state'] = f5().get_translate_port_state([name])
                    d['type'] = f5().get_type([name])
                    d['vlan'] = f5().get_vlan([name])
                    return d
                return lambda: detail(f5)
            elif a.sub2 == 'brief':
                def brief(f5):
                    d = {}
                    d['actual_hardware_acceleration'] = f5().get_actual_hardware_acceleration([name])
                    d['default_pool_name'] = f5().get_default_pool_name([name])
                    d['destination'] = f5().get_destination_v2([name])
                    d['enabled_state'] = f5().get_enabled_state([name])
                    d['object_status'] = f5().get_object_status([name])
                    d['persistence_profile'] = f5().get_persistence_profile([name])
                    d['profile'] = f5().get_profile([name])
                    d['protocol'] = f5().get_protocol([name])
                    d['translate_address_state'] = f5().get_translate_address_state([name])
                    d['translate_port_state'] = f5().get_translate_port_state([name])
                    d['type'] = f5().get_type([name])
                    return d
                return lambda: brief(f5)
            elif a.sub2 == 'status':
                return lambda: f5().get_object_status([name])

        elif a.sub == 'create':
            protocol = a.sub2 if a.sub2 else 'PROTOCOL_TCP'
            if a.sub3:
                profile = a.sub3
            elif protocol == 'PROTOCOL_UDP':
                profile = 'udp'
            else:
                profile = 'tcp'
            poolname = a.sub4
            netmask = a.sub5 if a.sub5 else '255.255.255.255'
            vserver = {
                'name': name,
                'address': ip,
                'port': port,
                'protocol': protocol,
            }
            resource = { 'type': 'RESOURCE_TYPE_POOL' }
            if poolname: resource['default_pool_name'] = poolname
            profile = {
                    'profile_context': 'PROFILE_CONTEXT_TYPE_ALL',
                    'profile_name': profile,
            }
            self.info("vserver:%s netmask:%s resource:%s, profile:%s"
                    % (vserver, netmask, resource, profile))
            def vserver_create():
                f5().create([vserver], [netmask], [resource], [[profile]])
                # Auto disable NAT and PAT if nPath
                if profile['profile_name'] == 'nPath':
                    f5().set_translate_address_state([name], ['STATE_DISABLED'])
                    f5().set_translate_port_state([name], ['STATE_DISABLED'])
            return lambda: vserver_create()

        elif a.sub == 'delete':
            return lambda: f5().delete_virtual_server([name])

        elif a.sub == 'set':
            if a.sub2 == 'pool':
                poolname = a.sub3
                return lambda: f5().set_default_pool_name([name], [poolname])
            elif a.sub2 == 'nat':
                if a.sub3 == 'disabled':
                    return lambda: f5().set_translate_address_state([name], ['STATE_DISABLED'])
                elif a.sub3 == 'enabled':
                    return lambda: f5().set_translate_address_state([name], ['STATE_ENABLED'])
            elif a.sub2 == 'pat':
                if a.sub3 == 'disabled':
                    return lambda: f5().set_translate_port_state([name], ['STATE_DISABLED'])
                elif a.sub3 == 'enabled':
                    return lambda: f5().set_translate_port_state([name], ['STATE_ENABLED'])
            elif a.sub2 == 'snat':
                if a.sub3 == 'none':
                    return lambda: f5().set_source_address_translation_none([name])



    def run(self):
        ''' Do the actual stuff.
            We are doning some lazy evaluation stuff here. The command line
            tool does not do anything with the slow F5 API until it is clear
            what to do and that there is no semantic or syntax error. '''

        a = self._args
        lazy = None

        if a.name:
          # Remove the /partition/ prefix, setting default partition after 
          # login instead
          a.name = re.sub(self._partition, '', a.name)
          a.name = re.sub('^/+', '', a.name)

        if a.what == 'node':
            lazy = self.__do_node(lambda: self._f5.LocalLB.NodeAddressV2)
        elif a.what == 'monitor':
            lazy = self.__do_monitor(lambda: self._f5.LocalLB.Monitor)
        elif a.what == 'pool':
            lazy = self.__do_pool(lambda: self._f5.LocalLB.Pool)
        elif a.what == 'vserver':
            lazy = self.__do_vserver(lambda: self._f5.LocalLB.VirtualServer)

        if isfunction(lazy):
            self.info('Doing some stuf via the API, it may take a while')
            self.__login()
            self.out(lazy())
        else:
            print_synopsis()
            sys.exit(1)



if __name__ == '__main__':
    ''' The main function, here we will have Popcorn for free! '''

    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument('-h', action='store_true', help='Help')
    parser.add_argument('-v', action='store_true', help='Verbose')
    parser.add_argument('-V', action='store_true', help='Print version')
    parser.add_argument('-m', action='store', help='The default lbmethod',
        default='LB_METHOD_ROUND_ROBIN')
    parser.add_argument('-C', action='store', help='Config file',
        default=expanduser('~') + '/.fapi.conf')

    parser.add_argument('what', nargs='?', help='node|pool|monitor|vserver|...')
    parser.add_argument('name', nargs='?', help='The object name to operate on')
    parser.add_argument('sub', nargs='?', help='First sub command')
    parser.add_argument('sub2', nargs='?', help='Second sub command')
    parser.add_argument('sub3', nargs='?', help='Third sub command')
    parser.add_argument('sub4', nargs='?', help='Fourth sub command')
    parser.add_argument('sub5', nargs='?', help='Fith sub command')

    args = parser.parse_args()

    if args.h:
        parser.print_help()
        print ''
        print_synopsis()
        sys.exit(0)

    if args.V:
        sys.exit(0)

    fapi = Fapi(args)

    fapi.run()

#   try: 
#       if not fapi.run():
#           fapi.info('Don\'t know what to do')
#           sys.exit(1)
#   except Exception, e:
#       fapi.info(e)
#       sys.exit(2)

# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
